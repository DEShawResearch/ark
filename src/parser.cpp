#include "exception.hpp"
#include "parser.hpp"

#include <sstream>
#include <fstream>
#include <cstring>
#include <cstdlib>
#include <climits>

#include <sys/stat.h>

namespace Ark {

  // static member initialization
  tokenizer::syntax parser::key_syn("{}<>[]!.=","#");
  tokenizer::syntax parser::val_syn("{}<>[]!?","#");
  tokenizer::syntax parser::no_syn("{}<>[]","#");

  inline
  std::string throwInputError(exception         &orig,
                              const std::string &msg,
                              const char        *file,
                              tokenizer         &t) {
    std::ostringstream o;
    o << orig.what() << "\n"
      << msg << " at " << (file?file:"???")
      << ":" << t.lineno() << ":" << t.colno();
    throw InputError(o.str());
  }

  //public:
  ark &parser::parse(ark &a,std::istream &in) const {
    tokenizer t(in,no_syn);
    try {
      t.next(val_syn);
      a = parse_value(t);
      if (t.next().kind() != token::End )
        throw InputError("extra stuff after the value");
    }
    catch (exception &e) {
      throwInputError(e,"parse problem",current_file,t);
    }
    return a;
  }
  ark &parser::parse(ark &a,const std::string &s) const {
    std::istringstream In(s);
    return parse(a,In);
  }
  //public:
  ark &parser::parse_keyvals(ark &a,std::istream &in) const {
    tokenizer t(in,no_syn);
    try {
#ifdef ANARKY_NOW
      ark::annotation_t anno(current_file,t.lineno(),t.colno());
      if (a.kind()!=Table) a.annotate(anno);
#endif
      a.be(Table);
      while( t.next(key_syn).kind()!=token::End ) parse_keyvalue(a,t);
    }
    catch (exception &e) {
      throwInputError(e,"parse problem",current_file,t);
    }
    return a;
  }
  ark &parser::parse_keyvals(ark &a,const std::string &s) const {
    std::istringstream In(s);
    return parse_keyvals(a,In);
  }
  //private:
  void parser::parse_keyvalue(ark &a,tokenizer &t) const {
    // check for !include and other key-like specials
    if (t.current().kind()==token::Syntax
        && t.current().syntax()=='!') {
      if (t.next().kind() != token::Symbol)
        throw InputError("expecting a special symbol");

      if (t.current().text()=="include") {
        t.next();
        if (t.current().kind()==token::Symbol
            || t.current().kind()==token::String) {
          try {
            include_file(a,t.current().text());
          }
          catch (exception &e) {
            throwInputError(e,"include problem",current_file,t);
          }
        }
        else
          throw InputError("!include expected a string or quoted string");
      }
      else
        throw InputError("unknown special token");
      return;
    } // nothing special

    // OK, I expect to see a key now
    if (t.current().kind()!=token::Symbol)
      throw InputError("expecting a key symbol");

#ifdef ANARKY_NOW
    ark::annotation_t anno(current_file,t.lineno(),t.colno());
    if (a.kind()!=Table) a.annotate(anno);
#endif

    table_t &d=a.be(Table).table(); // I must be a table
    key_t k=t.current().text();     // this must be my key
    t.next(key_syn);

    // possible special syntax for !erase
    if (t.current().kind()==token::Syntax
        && t.current().syntax()=='!') {
      if (t.next().kind() != token::Symbol)
        throw InputError("expecting a special symbol");

      if (t.current().text()=="erase") {
        d.erase(k);
      }
      else
        throw InputError("unknown special token");
      return;
    } // nothing special

    descend(d[k],t);
  }
  // private:
  void parser::descend(ark &a,tokenizer &t) const {
#ifdef ANARKY_NOW
    ark::annotation_t anno(current_file,t.lineno(),t.colno());
#endif
    if (t.current().syntax()=='[') {
      unsigned offset;
#ifdef ANARKY_NOW      
      if (a.kind()!=Vector) a.annotate(anno);
#endif
      // we have a vector access
      vector_t &v=a.be(Vector).vector();

      // read a number or "+"
      if (t.next(key_syn).kind()!=token::Symbol)
        throw InputError("expecting a number");
      if (t.current().text()=="+") {
        offset = v.size();
      }
      else {
        const char * S=t.current().text().c_str();
        char       * E=NULL;
        offset = strtoul(S,&E,10);
        if (S==E || *E) throw InputError("unable to parse strange number");
      }
      // read ']'
      if (t.next(key_syn).syntax()!=']') throw InputError("expecting ']'");

      if (offset==v.size()) v.push_back(ark(None));
      else if (offset > v.size())
        throw InputError("non-contiguous vector set not allowed");
      t.next(key_syn); // get next and keep descending

      // possible special syntax for !erase
      if (t.current().kind()==token::Syntax
          && t.current().syntax()=='!') {
        if (t.next().kind() != token::Symbol)
          throw InputError("expecting a special symbol");

        if (t.current().text()=="erase") {
          v.erase(v.begin()+offset);
        }
        else
          throw InputError("unknown special token");
        return;
      } // nothing special

      descend(v[offset],t);
    }
    // t now is on a . is more key or an = if value is next
    // now follow any array references after it
    else if (t.current().syntax()=='.') {
      t.next(key_syn); // back into a keyvals context
      parse_keyvalue(a,t);
    }
    else if (t.current().syntax()=='{') {
#ifdef ANARKY_NOW
      if (a.kind()!=Table) a.annotate(anno);
#endif
      // namespace
      a.be(Table);
      while( t.next(key_syn).syntax()!='}') parse_keyvalue(a,t);
    }
    else if (t.current().syntax()=='=') {
      // assignment
      t.next(val_syn);
      a = parse_value(t);
      //a.annotate(anno);
    }
    else throw InputError("expecting '.' or '=' or '{'");
  }
  namespace {
    std::string pathify(const std::string &s,const char *path) {
      if (!path || (s.size() && s[0]=='/')) return s;
      const char *ind = strrchr(path,'/');
      if (!ind) return s;
      return std::string(path,ind+1)+s;
    }
  }
  //private:
  ark parser::parse_value(tokenizer &t) const {
    ark a;
#ifdef ANARKY_NOW
    ark::annotation_t anno(current_file,t.lineno(),t.colno());
    a.annotate(anno);
#endif
    // check for special form
    if (t.current().syntax()=='!') {
      if (t.next().kind() != token::Symbol)
        throw InputError("expecting a special symbol");

      if (t.current().text()=="file") {
        t.next();
        if (t.current().kind()==token::Symbol
            || t.current().kind()==token::String) {
          a.be(Atom); // use current_file to get relative path
          a.atom() = pathify(t.current().text(),current_file);
        }
        else
          throw InputError("!file expected a string or quoted string");
      }
      else
        throw InputError("unknown special token");
    }
    else { // ok, not special
      switch(t.current().kind()) {
      case token::Symbol:
      case token::String:
        a.be(Atom);
        a.atom() = t.current().text();
        break;
      case token::Syntax:
        switch( t.current().syntax() ) {
        case '[':
          a.be(Vector);
          for(int i=0; t.next(val_syn).syntax()!=']'; i++)
            a.vector().push_back(parse_value(t));
          break;
        case '{':
          a.be(Table);
          while( t.next(key_syn).syntax()!='}') parse_keyvalue(a,t);
          break;
        case '?':
          a.be(None);
          break;
        default:
          throw InputError("expecting '{' or '[' or '?'");
        }
        break;
      default:
        throw InputError("expecting '{' or '[' or '?' or string");
      }
    }
    return a;
  }
  //private:
  void parser::include_file(ark &a,const std::string &f) const {
    if (f.empty())        throw InputError("include filename is empty");
    if (include_depth>20) throw InputError("include depth exceeded");

    // form path
    std::string path = pathify(f,current_file);

    parser tmp;
    tmp.current_file = path.c_str();
    tmp.include_depth = include_depth+1;

    try {
      std::ifstream F;
      // need this because ifstream will open a dir without complaining
      // and act as if it is an empty file.
      struct stat statbuf;
      if(stat(tmp.current_file, &statbuf) == -1) {
        std::ostringstream o;
        o << "unable to stat file: " << tmp.current_file;
        o << " (" << strerror(errno) << ")";
        throw InputError(o.str());
      }
      if (!S_ISREG(statbuf.st_mode)){
        std::ostringstream o;
        o << "not a regular file: " << tmp.current_file;
        throw InputError(o.str());
      }
      F.open(tmp.current_file);
      // ok, done with that mess
      if (F) tmp.parse_keyvals(a,F);
      else {
        std::ostringstream o;
        o << "unable to read file: " << tmp.current_file;
        throw InputError(o.str());
      }
    }
    catch (exception &e) {
      std::ostringstream o;
      o << "unable to parse file: " << tmp.current_file
        << "\n" << e.what();
      throw InputError(o.str());
    }
  }
}
