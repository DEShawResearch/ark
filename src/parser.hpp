#ifndef ark_parser_hpp
#define ark_parser_hpp

#include "base.hpp"
#include "tokens.hpp" // for my private members

/*! \file ark/parser.hpp

  The parser grammar is a superset of the strict grammar (again
  we don't have any standard for parsing boxes).  This superset was
  chosen to allow for shorthands which users would need to invoke from the
  command line, includes to allow online insertion of files, and
  overrides to allow some primitive inheritance/reuse.
  <code>
\verbatim
---PARSER GRAMMAR---
ARK  -> NONE | STRING | [ARK*] | { KEYVAL* }
NONE -> ?
STRING is a ", ', or ` delimited string with simple '\' meta-charactering
       or a bare string (no intervening white-space or grammar symbols).
KEYVAL  -> INCLUDE | SKEY=ARK | SKEY ARK
  the first form does assignment, the second form does an "enclosure",
  expecting ARK to be a table and then making the SKEYed element a
  table with additional keys taken from ARK.
INCLUDE -> ! include STRING
  expands the file named by STRING, parsing the contents as KEYVAL*
SKEY    -> IKEY | SKEY . IKEY
  keys which name tables can be subindexed with . in a familiar way
IKEY    -> KEY | IKEY [ INDEX ]
  keys which name vectors are indexed in similarly to arrays.
KEY is a non-delimited string in the regexp [_a-xA-Z][_a-zA-Z0-9]*
INDEX   -> + | INT
  if + the index is taken to be the end+1 element of the array, i.e.
  the assignment is taken to be a 'push_back' operation.
INT is a base 10 non-negative integer.
\endverbatim
  </code>
*/

namespace Ark {

/*! parser defines a helper type used for parsing arks.  In theory,
  having a helper type for the parser allows some user-level control
  over the parsing.  In practice, the parser type isn't that
  malleable, and just serves as a place to hold auxilliary parsing
  state which possibly could be done another way.

  Example:
<code>
\verbatim
  Ark::ark b;
  parser().parse_file(b,argv[1]);
\endverbatim
</code>

  This style has some similarities with ark::printer, but since member
  functions that modify the parsed syntax are strongly discouraged,
  it's not clear how far this similarity can or should be pushed.  An
  alternative parser is, for all intents and purposes, an alternative
  ark language, so unless one can justify "balkanizing" the ark
  language, alternative parsers will be rejected.

  A parser takes an ark to update.  It does not clear its input, but
  merely overrides it.  If this is not what the user wants, the user
  should call the ark::clear() member before using the parser.
*/
  class parser {
    unsigned include_depth;    //!< current include file depth.
    const char *current_file;  //!< name of current include file.

    ark parse_value(tokenizer &t) const;

    void descend(ark &a,tokenizer &t) const;
    void parse_keyvalue(ark &a,tokenizer &t) const;
    void include_file(ark &a,const std::string &f) const;

  public:
    static tokenizer::syntax key_syn; //!< punctuation for tokens in a key context.
    static tokenizer::syntax val_syn; //!< punctuation for tokens in a value context.
    static tokenizer::syntax no_syn;  //!< punctuation for tokens any other context.

    /*! Constructor.
      Sets the syntax and initializes the include file meta-information.
    */
    parser() : include_depth(0),current_file(NULL) {};
    /*! parse into an ark from a C++ stream.
      Starts from the production rule for ARK in the grammar.
      @param a input ark.
      @param in input stream.
      @return reference to a.
    */
    ark &parse(ark &a,std::istream &in) const;
    /*! parse() via stringstream.
      @param a input ark.
      @param s input string.
      @return reference to a.
    */
    ark &parse(ark &a,const std::string &s) const;

    /*! parse into a sequence of key-value pairs from a C++ stream.
      Useful for reading config files.
      Starts from the production rule for KEYVAL* in the grammar.
      @param a input ark.
      @param in input stream.
      @return reference to a.
    */
    ark &parse_keyvals(ark &a,std::istream &in) const;
    /*! parse_keyvals() via stringstream.
      @param a input ark.
      @param s input string.
      @return reference to a.
    */
    ark &parse_keyvals(ark &a,const std::string &s) const;

    /*! parse the contents of a file as a sequence of key-value pairs.
      Useful for reading config files.
      Starts from the production rule for KEYVAL* in the grammar.
      Checks include depth and updated the include file meta-information.
      @param a input ark.
      @param s file to read.
      @return reference to a.
    */
    ark &parse_file(ark &a,const std::string &s) const {
      include_file(a,s);
      return a;
    }
  };
}



#endif
