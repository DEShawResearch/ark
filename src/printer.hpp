#ifndef ark_printer_hpp
#define ark_printer_hpp

#include "base.hpp"

#include <cstdio>
#include <ostream>

/*! \file ark/printer.hpp

Although the ark class has an ostream output method, a variety of
modifications to the output 'style' might be warranted, such as the
addition of white-space for pretty printing or the removal of outer
delimiters for 'config file' generation.  Maybe you'd like to write to
a C FILE*.  To support these modified forms, we define a printer class
which interacts in nice ways with C++ iostreams and C FILE*.

Example:
<code>
\verbatim
    Ark::ark a;
    Ark::printer P;
    P.no_delim(1).whitespace(1);
    cout << P(a) << endl;
    cout << P.no_delim(0)(a) << endl;
\endverbatim
</code>
*/

namespace Ark {
  /*! Consider this a 'debug only' sort of output which isn't intended
    to look very good.
    @param f C-style stream.
    @param a output ark.
  */
  void fdump(FILE *f,const ark &a);

  //! The collection of output modification flags currently recognized.
  struct print_flags {
    bool  no_delim;   //!< omit delimiters of outermost ark element
    bool  whitespace; //!< insert whitespace for pretty printing.
    bool  open_tables;//!< print tables as key {...} rather than key = {...}
    bool  flatten;    //!< display as extended-key = value pairs.
    unsigned width;   //!< start wrapping lines once this width is exceeded
    unsigned indent;  //!< number of spaces per indent
    //! default initializes everything to false.
    print_flags() : no_delim(0),whitespace(0),open_tables(0),flatten(0),
                    width(0),indent(4) {}
  };

  //! A printer is thought of as "something that knows how to print arks".
  class printer {
    print_flags flags; //!< current set of printing flags.

  public:
    //! A printer generates a print_ref when given an ark.
    //! printer_refs get sent to ostreams or take FILE*s.
    class printer_ref {
      print_flags flags; //! flags of the printer.
      const ark &a; //! the ark to print.
      const bool from_python;

    private:
      void output(const ark &a,std::ostream &o,
                  unsigned &ind,unsigned &col,bool top) const;
      void output_flatten(const std::string &k,const ark &a,std::ostream &o,
                          unsigned ind) const;

    public:
      //! Construct from flags and ark.
      //! @param f flags to print with.
      //! @param A ark to print.
      printer_ref(const print_flags &f,const ark &A, bool from_python=false) : flags(f), a(A), from_python(from_python) {}
      //! public function to be called by operator<<().
      //! @param o output.
      void output(std::ostream &o) const { 
        unsigned i=0,c=0;
        output(a,o,i,c,true);
      }
      //! public function to be called on C-style streams.
      //! @param f C-style stream.
      void fprint(FILE *f) const;
    };

    //! trivial default constructor.
    printer() {}

    /*! chain-able setter for width.
      @param b new value for width flag.
      @return reference to this printer.
    */
    printer &width(unsigned b) { flags.width = b; return *this; }

    /*! chain-able setter for indent.
      @param b new value for indent flag.
      @return reference to this printer.
    */
    printer &indent(unsigned b) { flags.indent = b; return *this; }

    /*! chain-able setter for whitespace.
      @param b new value for whitespace flag.
      @return reference to this printer.
    */
    printer &whitespace(bool b) { flags.whitespace = b; return *this; }

    /*! chain-able setter for no_delim.
      @param b new value for no_delim flag.
      @return reference to this printer.
    */
    printer &no_delim(bool b) { flags.no_delim = b; return *this; }

    /*! chain-able setter for open_tables
      @param b new value for open_tables flag.
      @return reference to this printer.
    */
    printer &open_tables(bool b) { flags.open_tables = b; return *this; }

    /*! chain-able setter for flatten
      @param b new value for flatten flag.
      @return reference to this printer.
    */
    printer &flatten(bool b) { flags.flatten = b; return *this; }

    /*! Apply printer to an ark and give a printer reference.
      @param a the ark to print.
      @return a printer_ref built from a and flags.
    */
    printer_ref operator()(const ark &a, bool from_python=false) const {
      return printer_ref(flags,a,from_python);
    }
  };
}

/*! C++-stream output (calls printer_ref::output()).
  @param o output stream.
  @param x printer reference.
  @return o
*/
std::ostream &operator<<(std::ostream &o,const Ark::printer::printer_ref &x);

#endif
