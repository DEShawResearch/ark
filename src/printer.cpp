#include "printer.hpp"
#include "parser.hpp" // for syntax definitions

#include <sstream>
#include <climits>

namespace Ark {

  static bool requires_quotes(atom_t const& a) {
    if (a.c_str()[0] == '\0') return true;
    tokenizer::syntax const& syn = parser::val_syn;
    for (const char* c = a.c_str(); *c; ++c) {
      if (syn.is_reserved(*c) || isspace(*c) || *c=='\\') return true;
    }
    return false;
  }

  void fdump(FILE *f,const ark &a) {
    switch(a.kind()) {
    case None:
      fprintf(f,"?");
      return;
    case Atom:
      fprintf(f,"\"%s\"",a.atom().c_str());
      return;
    case Vector: {
      fprintf(f,"[");
      vector_t::const_iterator i;
      for(i=a.vector().begin(); i != a.vector().end(); ++i)
	fdump(f,*i);
      fprintf(f,"]");
      return;
    }
    case Table: {
      fprintf(f,"{");
      table_t::const_iterator i;
      for(i=a.table().begin(); i != a.table().end(); ++i) {
	fprintf(f,"%s=",i->first.c_str());
	fdump(f,i->second);
      }
      fprintf(f,"}");
      return;
    }
    }
  }

  namespace {
    void space(std::ostream &o,unsigned sp) {
      for(unsigned j=0; j < sp; ++j) o<<" ";
    }
  }

  void printer::printer_ref::output_flatten(const std::string &key,
                                               const ark &a,
                                               std::ostream &o,
                                               unsigned ind) const {
    const bool whitespace = flags.whitespace;
    const unsigned tab = flags.indent;
    if (a.kind()==None || a.kind()==Atom ||
        (a.kind()==Vector && a.vector().empty()) ||
        (a.kind()==Table && a.table().empty()) ) {

      if (whitespace) space(o,tab*ind);

      o<<key;
      if (whitespace) o<<" = ";
      else            o<<"=";

      switch(a.kind()) {
      case None: (o<<"?"); break;
      case Atom: {
        std::ostringstream oss;
        const char* s = a.atom().c_str();
        if (from_python && (unsigned char)(*s) == 0xff) {
            oss << "!file ";
            ++s;
        }
        const bool with_quotes = !whitespace || requires_quotes(a.atom());
        if (with_quotes) oss<<"\"";
        for(; *s ; oss<<*s, ++s)
          if (*s=='\\' || *s=='"') oss<<"\\";
        if (with_quotes) oss<<"\"";
        o<<oss.str();
      } break;
      case Vector: (o<<"[]"); break;
      case Table: (o<<"{}"); break;
      }

      if (whitespace) o<<"\n";
    }
    else if (a.kind()==Vector) {
      for(size_t i=0; i < a.vector().size(); ++i) {
        std::stringstream sst;
        sst << key << "[" << i << "]";
        output_flatten(sst.str(), a.vector()[i], o, ind);
      }
    }
    else if (a.kind()==Table) {
      for(table_t::const_iterator p=a.table().begin();
          p != a.table().end(); ++p)
        output_flatten(key + "." + p->first.str(),p->second,o,ind);
    }
    else {
      // THROW?
    }
  }

  void printer::printer_ref::output(const ark &a,
                                    std::ostream &o,
                                    unsigned &ind,unsigned &col,
                                    bool top) const {
    const bool delim = !(flags.no_delim && top);
    const bool whitespace = flags.whitespace;
    const bool open_tables = flags.open_tables;
    const bool flatten = flags.flatten;
    const unsigned width = flags.width?flags.width:INT_MAX;
    const unsigned tab = flags.indent;

    switch(a.kind()) {
    case None:
      (o<<"?");
      break;
    case Atom:
      if (delim) {
        std::ostringstream oss;
        const char* s = a.atom().c_str();
        if (from_python && (unsigned char)(*s) == 0xff) {
            oss << "!file ";
            ++s;
        }
        const bool with_quotes = !whitespace || requires_quotes(a.atom());
        if (with_quotes) oss<<"\"";
        for(; *s ; oss<<*s, ++s)
          if (*s=='\\' || *s=='"') oss<<"\\";
        if (with_quotes) oss<<"\"";
        o<<oss.str(), col+=oss.str().size();
      }
      else { // no-delimited atoms don't escape
        const char* s = a.atom().c_str();
        if (from_python && (unsigned char)(*s) == 0xff) {
            o << "!file ";
            ++s;
        }
        o << s;
        col += a.atom().str().size() + 5;
      }
      break;
    case Vector: {
      if (delim) o<<"[", ++col;
      unsigned rem=col;
      for(vector_t::const_iterator i=a.vector().begin();
          i != a.vector().end(); ++i) {

        if (whitespace &&
            i!=a.vector().begin()) {
          if (col > width) o<<"\n", col=rem, space(o,col);
          else o<<" ", ++col;
        }

        output(*i,o,ind,col,false);
      }
      if (delim) o<<"]", ++col;
    } break;
    case Table: {
      if (delim) {
        ++ind;
        o<<"{",++col;
        if (whitespace) o<<"\n";
      }
      for(table_t::const_iterator p=a.table().begin();
          p != a.table().end(); ++p) {
        if (flatten) {
          output_flatten(p->first.str(),p->second,o,ind);
        }
        else {
          if (whitespace) col = tab*ind, space(o,col);

          o<<p->first.str(), col+=p->first.str().size();
          if( open_tables && p->second.kind() == Table ) {
            if (whitespace) o<<" ", ++col;
            else /* pass */ ;
          }
          else {
            if (whitespace) o<<" = ", col+=3;
            else            o<<"=", ++col;
          }
          output(p->second,o,ind,col,false);
          
          if (whitespace) o<<"\n";
        }
      }
      if (delim) {
        --ind;
        if (whitespace) col = tab*ind, space(o,col);
        o<<"}",++col;
      }
    } break;
    }
  }

  namespace { // simple FILE* to ostream adapter
    struct FILEbuf : std::basic_streambuf< char, std::char_traits< char > > {
      FILE *_f;
      FILEbuf(FILE *f) : _f(f) {}

      // For EOF detection above
      typedef std::char_traits< char > Traits;

      // This is called every time characters are put to stream.
      int overflow(int c = Traits::eof()) {
        // Handle output
        if (c != Traits::eof()) fputc(c,_f);
        // I'm not sure about this return value!
        return 0;
      }
      
      // This function is called when stream is flushed,
      // for example when std::endl is put to stream.
      int sync(void) { return fflush(_f); }
    };

    struct FILEstream : std::basic_ostream< char, std::char_traits< char > > {
      FILEbuf buf;
      FILEstream(FILE *f) :
        std::basic_ostream< char, std::char_traits< char > >(&buf),buf(f) {}
    };
  }

  void printer::printer_ref::fprint(FILE *f) const {
    FILEstream fs(f);
    output(fs);
  }

}

std::ostream &operator<<(std::ostream &o,const Ark::printer::printer_ref &x) {
  x.output(o); return o;
}
