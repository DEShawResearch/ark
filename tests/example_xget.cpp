#include <ark/parser.hpp>
#include <ark/printer.hpp>

#include <cstdlib>
#include <iostream>
#include <sstream>

using namespace Ark;
using namespace std;

namespace Ark { // really, if you want to do your checking don't pollute base

  std::string xgetstr(const ark &a,const std::string &q) {
    const ark * b = a.xget(q);
    if (!b)
      throw exception("ark_xgetstr: key: " + q + " not found");
    const atom_t * c = b->get_atom();
    if (!c)
      throw exception("ark_xgetstr: key: " + q + " not an atom");
    return c->str();
  }

}

std::string annotation(const ark &a) {
#ifdef ANARKY_NOW
  std::ostringstream ss;
  ss << "<" << a.annotation().file << ":" << a.annotation().lineno << "/" << a.annotation().colno << ">";
  return ss.str();
#else
  return "";
#endif
}

int main(int argc,char **argv) {

  ark a;

  if (argv[1]) {
    try { parser().parse_file(a,argv[1]); }
    catch (...) {
      fprintf(stderr,"parse failed\n");
      exit(1);
    }
  }
  else {
    cerr << "usage: " << argv[0] << " testfile" << endl;
    exit(1);
  }
  

  const ark *t;
  printer P;
  P.whitespace(1);

  cout << P(a) << endl;
  cout << "#####queries#####" << endl;

  if ( (t = a.xget("ggg")) )
    cout << "ggg" << " is " << P(*t) << annotation(*t) << endl;

  if ( (t = a.xget("ggg.rrr")->get(1)) )
    cout << "ggg.rrr[1]" << " is " << P(*t) << annotation(*t) << endl;

  if ( (t = a.xget("ggg.rrr")->get(2)) )
    cout << "ggg.rrr[2]" << " is " << P(*t) << annotation(*t) << endl;

  // Or xgetstr
  try {
      cout << "ggg.rrr[2]" << " is " << xgetstr(a,"ggg.rrr[2]") << endl;
      cout << "ggg.rrr[1]" << " is " << xgetstr(a,"ggg.rrr[1]") << endl; // throws with ./testfile
  }catch(Ark::exception &e){
      cout << "Caught: " << e.what() << "\n";
  }

  // misspelled
  if ( (t = a.xget("ggg.rr[1]")) )
    cout << "ggg.rr[1]" << " is " << P(*t) << annotation(*t) << endl;

  if ( (t = a.xget("x[1].tau")) )
    cout << "x[1].tau" << " is " << P(*t) << annotation(*t) << endl;

  return 0;
}
