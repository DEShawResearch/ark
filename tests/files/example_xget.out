{A={x="1"
    y="2"}
 B={y="2"}
 array=["1"
        "2"
        "3"
        "0x27"
        "010"]
 ert="17"
 f=["1"]
 file="blah blah `file` here's \\`'s"
 foo=["3"]
 ggg={gotincluded=["1"
                   "2"
                   "3"]
      rrr=["5"
           {jj="183"}
           "xxx"]}
 jgak="123"
 nothing=["1"
          ?
          "4"]
 t={e="4"
    multi={gotincluded=["1"
                        "2"
                        "3"]
           ijkl="abcd"
           record="here"}}
 x=[{tau=["1"
          "2"]}
    {tau=["23"
          {hey="there"}
          "56"]}]
 y=[{myarray=["3"
              "4"
              "5"]}
    "yuiop"
    "qwerty"
    "added"
    "added again"]
 z="5"}
#####queries#####
ggg is {gotincluded=["1"
              "2"
              "3"]
 rrr=["5"
      {jj="183"}
      "xxx"]}
ggg.rrr[1] is {jj="183"}
ggg.rrr[2] is "xxx"
ggg.rrr[2] is xxx
Caught: Exception: Ark::exception
Context:
xgetstr:  key: ggg.rrr[1] is not an Atom @/usr/include/ark/base.hpp:326 in std::string Ark::ark::xgetstr(const std::string&) const

x[1].tau is ["23"
 {hey="there"}
 "56"]
